### PR TITLE
Support batched QR decomposition

### PR DESCRIPTION
**Pull request by mrfh92**
## Due Diligence

- General:
    - [X]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [ ] ~~benchmarks: created for new functionality~~
    - [X] (within the error margin) benchmarks: performance improved or maintained 
    - [x] documentation updated where needed

## Description

Batched QR 

Issue/s resolved: necessary for #1696 

## Changes proposed:

ht.linalg.qr now is able to compute QR for batches of matrices, as PyTorch's batched linear algebra does

## Type of change
(not really) new feature 

#### Does this change modify the behaviour of other functions? If so, which?
yes (ht.linalg.qr)

### CODE REVIEW

**Comment by JuanPedroGHM** [comment]
### Benchmarks results - Sponsored by [perun](https://github.com/Helmholtz-AI-Energy/perun)

| function                                |   mpi_ranks | device   | metric   |      value |   ref_value |         std |   % change | type            | alert   |   lower_quantile |   upper_quantile |
|-----------------------------------------|-------------|----------|----------|------------|-------------|-------------|------------|-----------------|---------|------------------|------------------|
| matmul_split_0                          |           4 | CPU      | RUNTIME  | 0.13726    |  0.181238   | 0.0256314   |  -24.265   | jump-detection  | True    |      nan         |      nan         |
| apply_inplace_normalizer                |           4 | CPU      | RUNTIME  | 0.00320091 |  0.00104532 | 0.00649931  |  206.213   | jump-detection  | True    |      nan         |      nan         |
| matmul_split_0                          |           4 | GPU      | RUNTIME  | 0.033585   |  0.0562632  | 0.0285393   |  -40.3074  | jump-detection  | True    |      nan         |      nan         |
| matmul_split_1                          |           4 | GPU      | RUNTIME  | 0.0259047  |  0.0343244  | 0.0123812   |  -24.5298  | jump-detection  | True    |      nan         |      nan         |
| concatenate                             |           4 | GPU      | RUNTIME  | 0.173058   |  0.142046   | 0.0596402   |   21.8327  | jump-detection  | True    |      nan         |      nan         |
| qr_split_1                              |           4 | CPU      | RUNTIME  | 0.197579   |  0.186863   | 0.0262833   |    5.73458 | trend-deviation | True    |        0.179407  |        0.196056  |
| hierachical_svd_tol                     |           4 | CPU      | RUNTIME  | 0.0513506  |  0.0521396  | 0.000182696 |   -1.51317 | trend-deviation | True    |        0.0515701 |        0.0531136 |
| reshape                                 |           4 | CPU      | RUNTIME  | 0.155652   |  0.160234   | 0.00165965  |   -2.85945 | trend-deviation | True    |        0.156578  |        0.167456  |
| resplit                                 |           4 | CPU      | RUNTIME  | 1.07068    |  1.0918     | 0.00795151  |   -1.93423 | trend-deviation | True    |        1.0711    |        1.11149   |
| apply_inplace_robust_scaler_and_inverse |           4 | CPU      | RUNTIME  | 2.36339    |  2.45182    | 0.0576843   |   -3.60671 | trend-deviation | True    |        2.38057   |        2.56004   |
| matmul_split_0                          |           4 | GPU      | RUNTIME  | 0.033585   |  0.0501268  | 0.0285393   |  -33       | trend-deviation | True    |        0.0339485 |        0.0687051 |
| lanczos                                 |           4 | GPU      | RUNTIME  | 0.618193   |  0.595779   | 0.00583022  |    3.76207 | trend-deviation | True    |        0.581043  |        0.617789  |
| hierachical_svd_tol                     |           4 | GPU      | RUNTIME  | 0.123252   |  0.119621   | 0.000163931 |    3.03512 | trend-deviation | True    |        0.117207  |        0.122445  |
| kmeans                                  |           4 | GPU      | RUNTIME  | 0.680437   |  0.646702   | 0.00515695  |    5.21645 | trend-deviation | True    |        0.623644  |        0.677172  |
| apply_inplace_robust_scaler_and_inverse |           4 | GPU      | RUNTIME  | 6.21877    |  5.8251     | 0.0352397   |    6.75803 | trend-deviation | True    |        5.51667   |        6.17805   |

[Grafana Dashboard](http://57bc8d92-72f2-4869-accd-435ec06365cb.ka.bw-cloud-instance.org:3000/d/adjpqduq9r7k0a/heat-cb?orgId=1)
Last updated: 2024-11-15T08:33:51Z

File: heat/core/linalg/qr.py [hunk 1]
```diff
@@ -87,65 +90,57 @@ def qr(
...
         procs_to_merge = A.comm.size
 
-    if A.ndim != 2:
-        raise ValueError(
-            f"Array 'A' must be 2 dimensional, buts has {A.ndim} dimensions. \n Please open an issue on GitHub if you require QR for batches of matrices similar to PyTorch."
-        )
+    # if A.ndim != 2:
+    #     raise ValueError(
+    #         f"Array 'A' must be 2 dimensional, buts has {A.ndim} dimensions. \n Please open an issue on GitHub if you require QR for batches of matrices similar to PyTorch."
+    #     )
```
Comment (ClaudiaComito) [inline heat/core/linalg/qr.py:96]: dead code?

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Thanks a lot @mrfh92! I only have 2 tiny changes, otherwise it's ready as far as I can tell. 👍🏼

File: heat/core/linalg/qr.py (hunk 1 above)
Comment (mrfh92) [reply heat/core/linalg/qr.py]: removed

**Comment by mrfh92** [comment]
@JuanPedroGHM looks like the BM failed because the gitlab runner had to wait more than 01:30h for getting the job allocated. Is the cluster currently busy or out-of-order?

**Comment by mrfh92** [comment]
@ClaudiaComito I turned off the benchmark_pr as it does not work at the moment; except of that everything should be fine now

**Review by ClaudiaComito — APPROVED** [review]
Thanks again @mrfh92 !

**Pull request merged**